builder/src/lib.rs

```diff
@@ -1,10 +1,247 @@
+                        Some(b) => b,
+                        None => return TokenStream::from(quote! {
+                            compile_error!("missing or invalid `each` parameter in builder");
+                        }),
+                    };
+                    let optional_fields = get_optional_fields(&fields);
+                    (
+                        expand_field_definitions(&fields, &optional_fields, &builders),
+                        expand_field_initializers(&fields, &builders),
+                        expand_field_setters(&fields, &optional_fields, &builders),
+                        expand_build(&name, &fields, &optional_fields, &builders),
+                    )
+                }
+                _ => { (quote!{}, quote!{}, quote!{}, quote!{}) },
+            }
+        },
+        _ => unimplemented!(),
+    };
+
+    let expanded = quote! {
+        impl #name {
+            pub fn builder() -> #buildername {
+                #buildername { #fields_init }
+            }
+        }
+
+        pub struct #buildername {
+            #fields
+        }
+
+        impl #buildername {
+            #setters
+
+            pub fn build(&mut self) -> std::result::Result<#name, std::boxed::Box<dyn std::error::Error>> {
+                #build
+            }
+        }
+    };
+    TokenStream::from(expanded)
+}
+
+fn expand_field_definitions(fields: &syn::FieldsNamed, optional_fields: &HashSet<String>, builders: &HashMap<String, String>) -> TokenStream2 {
+    let f = fields.named.iter().map(|f| {
+        let ident = &f.ident;
+        let ty = &f.ty;
+        let identstr = f.ident.as_ref().map(|x| format!("{}", x)).unwrap_or("".to_owned());
+        if optional_fields.contains(&identstr) || builders.contains_key(&identstr) {
+            quote! {
+                #ident: #ty
+            }
+        } else {
+            quote! {
+                #ident: std::option::Option<#ty>
+            }
+        }
+    });
+    quote! { #(#f,)* }
+}
+
+fn expand_field_initializers(fields: &syn::FieldsNamed, builders: &HashMap<String, String>) -> TokenStream2 {
+    let f1 = fields.named.iter().filter(|f| {
+        !builders.contains_key(&f.ident.as_ref().unwrap().to_string())
+    }).map(|f| {
+        let ident = &f.ident;
+        quote! {
+            #ident: std::option::Option::None
+        }
+    });
+    let f2 = builders.keys().map(|k| {
+        // this should be def_site, but I'm unable to get the config to enable that method
+        let id = Ident::new(&k, proc_macro2::Span::call_site());
+        quote! { #id: Vec::new() }
+    });
+    quote! { #(#f1,)* #(#f2,)* }
+}
+
+fn expand_field_setters(fields: &syn::FieldsNamed, optional_fields: &HashSet<String>, builders: &HashMap<String, String>) -> TokenStream2 {
+    let setters = fields.named.iter().map(|f| {
+        let ident = &f.ident;
+        let ty = &f.ty;
+        let identstr = f.ident.as_ref().map(|x| format!("{}", x)).unwrap_or("".to_owned());
```

Comment:
On this line it's fine to unwrap the Option<Ident> since we got it from a syn::FieldsNamed which would be guaranteed to have named fields. I would write a plain unwrap() rather than having the fallback to "".